src/nvmet.c

```diff
@@ -833,209 +909,230 @@ int nvmet_revalidate_size(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char
+		} else if (rc == 1) {
+			print_error("NQN %s does not exists!", return_message, nqn);
+			free_nvmet_linked_lists(orig_ports, NULL);
+			return INVALID_VALUE;
+		}
 
-	list = clean_scandir(list, err);
+		/* Reset rc to INVALID_VALUE after port check */
+		rc = INVALID_VALUE;
 
-	if (err > 1) {
-		msg = "An invalid number of namespaces not created by RapidDisk exist.";
-		print_error("%s", return_message, msg);
-		return rc;
+		free_nvmet_linked_lists(orig_ports, NULL);
+	} else {
+		profile = nvmet_scan_subsystem(return_message, &rv);
+		if (rv == INVALID_VALUE) {
+			return rc;
+		}
+		orig_profile = profile;
+		while (profile != NULL) {
+			if (strcmp(nqn, profile->nqn) == 0) {
+				rc = SUCCESS;
+				break;
+			}
+			profile = profile->next;
+		}
+		if (rc != SUCCESS) {
+			print_error("NQN %s does not exists!", return_message, nqn);
+			free_nvmet_linked_lists(NULL, orig_profile);
+			return rc;
+		}
+
+		/* Reset rc to INVALID_VALUE after nqn check */
+		rc = INVALID_VALUE;
+
+		free_nvmet_linked_lists(NULL, orig_profile);
 	}
 
-	/*
-	 * Grab number of entries in allowed_hosts directory. We need this later, to check if one of them has been removed.
+
+	/* Example of this path:
+	 * /sys/kernel/config/nvmet/subsystems/nqn.2021-06.org.rapiddisk:ubuserver-rd0/allowed_hosts
 	 */
-	sprintf(path, "%s/%s%s-%s/allowed_hosts/", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
+	sprintf(path, "%s/%s/allowed_hosts", SYS_NVMET_TGT, nqn);
 	if ((allowed_host_number = scandir(path, &list, scandir_filter_no_dot, NULL)) < 0) {
-		msg = "Error. Unable to scan host exports directory for %s. %s: unlink: %s";
+		msg = "Error. Unable to access %s. %s: scandir: %s";
 		print_error(msg, return_message, path, __func__, strerror(errno));
 		return rc;
 	}
+
+	if (allowed_host_number > 0) {
+		for (n = 0; n < allowed_host_number; n++) {
+			strcat(target_string, list[n]->d_name);
+			strcat(target_string, " ");
+		}
+		size_t size = strlen(target_string);
+		size = size - 1;
+		target_string[size] = '\0';
+	}
+
 	list = clean_scandir(list, allowed_host_number);
 
-	/* Remove host NQN */
-	if (strlen(host) != 0) {
-		sprintf(path, "%s/%s%s-%s/allowed_hosts/%s", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device, host);
+	if (allowed_host_number > 0 && strlen(host) <= 0) {
+		msg = "Error. No host provided, but one or more target(s) is/are still mapped to NQN(s): %s. Please remove them first.";
+		print_error(msg, return_message, target_string);
+		return rc;
+	} else if (allowed_host_number > 0 && strlen(host) > 0) {
+		/* Example of this path:
+	 	 * /sys/kernel/config/nvmet/subsystems/nqn.2021-06.org.rapiddisk:ubuserver-rd0/allowed_hosts/<HOST>
+	 	 */
+		sprintf(path, "%s/%s/allowed_hosts/%s", SYS_NVMET_TGT, nqn, host);
 		if (access(path, F_OK) == SUCCESS) {
-			rc = unlink(path);
-			if (rc != SUCCESS) {
-				msg = "Error. Unable to remove host. %s: unlink: %s";
-				print_error(msg, return_message, __func__, strerror(errno));
+			if (unlink(path) != SUCCESS) {
```

Comment:
@matteotenca What if another NVMe Target is mapped to this host we are trying to remove. e.g. rd0 AND rd1 or rd2? Here we could be focusing on rd0 BUT we are assuming that no other /sys/kernel/config/nvmet/subsystems/<NQN>/allowed_hosts/<HOST> exists. Thoughts?